### ISSUE TITLE
Low performance with many small files

### ISSUE THREAD

**Issue by brauliobo** [minio/mc#2407]
The average speed of `mc mirror` of two hosts based on the AWS region is about 10kb/s when dealing with many small files in a bucket. Could maybe threads be used to parallelize the copy?

```
[ec2-user@ip-10-0-0-117 ~]$ ./mc mirror minio-old/opener-language-identifier minio-new/opener-language-identifier                                                                            
...4c41dc3d5f86fec/364551608.xml:  2.02 MB / 2.02 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 11.52 KB/s
```

**Comment by kannappanr** [comment]
@brauliobo Your issue seems similar to https://github.com/minio/mc/issues/2404 . Are you seeing a lot of time difference between uploading a single file that is of size lets say 2 MB vs multiple small files whose combined file size is 2MB. Small files will take a little more time, because of all the other extra things involved in listing the objects etc. But the difference should not be much. 

Can you please report the time taken in both the scenarios?

**Comment by systemlevel** [comment]
I'm having a similar issue with slow performance when moving many small photo files.  I've experienced this both with `mc cp` and `mc mirror` between minio and s3-compatible stores and between minio and local network stores.

<img width="728" alt="screen shot 2018-07-25 at 8 23 53 pm" src="https://user-images.githubusercontent.com/894983/43238115-043a244c-904a-11e8-844d-89718126d687.png">

<img width="745" alt="screen shot 2018-07-25 at 8 23 37 pm" src="https://user-images.githubusercontent.com/894983/43238116-044b0460-904a-11e8-8194-0b3913193e61.png">

At this rate, with the amount of small files we have in there this will take weeks to a month to complete.  Is there something I'm missing when it comes to moving many files with minio? I'm having a hard time identify what I might be doing incorrectly.  

We're on a fast 10Gig network here.

**Comment by systemlevel** [comment]
Quick additional test to this issue.  I tested going from a Google Cloud Compute instance (large instance, lots of RAM, CPU, etc. and network speeds of up ~900MB/s with speedtest-cli).

An `mc mirror` operation from the Google Cloud Compute instance to Wasabi resulted in the following transfer performance:

<img width="1162" alt="screen shot 2018-07-28 at 10 59 22 am" src="https://user-images.githubusercontent.com/894983/43358857-a2ecadaa-9255-11e8-969b-e094b1726800.png">

Just wanted to make sure it wasn't something with our own hardware/network.

**Comment by systemlevel** [comment]
Two week followup.  I've been running an mc mirror instance for 2 weeks now and it's still running. I'm still attempting to mirror all the files in my minio bucket to another S3 compatible:

![screen shot 2018-08-09 at 11 42 55 am](https://user-images.githubusercontent.com/894983/43916166-0972034a-9bca-11e8-810f-fa08e44d2732.png)

**Comment by kannappanr** [comment]
@systemlevel Sorry, for the late response. We have internally discussed a way to increase the number of parallel uploads for small files. I will be doing some tests and make sure that it works as intended. 

Just a bit of note about `mirror`, we basically move the objects from the `source` to the machine where you are running `mc` and then to the `target`. So, the speed might be half there.

**Comment by kannappanr** [comment]
@systemlevel We have started making few improvements and we are still testing this out. The initial fix is focussed on the progress bar not blocking threads that do the uploads. This should give you improvements in multiples of what you are seeing currently. We are also testing with the number of threads that we can have initially, once we optimize that, you should see even bigger improvements.

We have to make sure that we optimize that for all sizes of files and just not focus on small files.

Here is the PR that has the initial fix. https://github.com/minio/mc/pull/2521
You can checkout this PR and build.  If you want us to build, I can do it. Please let us know.

**Comment by systemlevel** [comment]
@kannappanr Is there a way to get the new build over brew for Mac? Perhaps there a tentative timeline for when that might be available? The other machine is a server running Ubuntu. What would you recommend for best way to build and reinstall mc client in these environments?  

I'd love to help test out the proposed update with our very large minio store.

**Comment by harshavardhana** [comment]
> @kannappanr Is there a way to get the new build over brew for Mac? Perhaps there a tentative timeline for when that might be available? The other machine is a server running Ubuntu. What would you recommend for best way to build and reinstall mc client in these environments?
> 
> I'd love to help test out the proposed update with our very large minio store.

@systemlevel here is a test binary for Linux [mc.gz](https://github.com/minio/mc/files/2320443/mc.gz), its gzipped `gunzip` it locally. 

When using `mc mirror` make sure you specify the desired parallelism using `mc mirror --worker-count=128`, worker counts maximum is 128.

**Comment by systemlevel** [comment]
I'm seeing an improvement on performance. My speed has more or less doubled with maxed out worker counts at 128.  

Previous average: ~500KB/s
New average: ~1.15MB/s

Here are some early results from the new binary:

<img width="1238" alt="screen shot 2018-08-25 at 12 36 48 pm" src="https://user-images.githubusercontent.com/894983/44621491-6933d280-a864-11e8-9a01-c3a5c8d68574.png">

Minio server working hard:

<img width="1479" alt="screen shot 2018-08-25 at 12 35 20 pm" src="https://user-images.githubusercontent.com/894983/44621500-836db080-a864-11e8-9167-540def27cb18.png">

Does this version of the binary include the following from @kannappanr? "We are also testing with the number of threads that we can have initially, once we optimize that, you should see even bigger improvements."

**Comment by harshavardhana** [comment]
> Does this version of the binary include the following from @kannappanr? "We are also testing with the number of threads that we can have initially, once we optimize that, you should see even bigger improvements."

Yes that is already there

> I'm assuming it's dropping a bit during the time when the minio server is attempting to see which files already exist on the target store? With these very large minio stores full of small photo files that seems to be when the speeds drops down a bit understandably I suppose.

yes they do

**Comment by systemlevel** [comment]
I'm getting slightly better performance when I set the workers down to 32:

~1.45MB/s

<img width="1251" alt="screen shot 2018-08-25 at 3 39 24 pm" src="https://user-images.githubusercontent.com/894983/44622779-4a8e0580-a87d-11e8-9813-809aa07ad99c.png">

When it was set to 128 I also experience a gateway timeout (shown in screenshot).

**Comment by harshavardhana** [comment]
> When it was set to 128 I also experienced a gateway timeout (shown in screenshot).

yes it is possible that 128 is overwhelming @systemlevel 32 is usually a good value, making equal to the number of CPU's is better. Also, you should not look at progress bar since it might be a bit misleading in numbers due to how it is updated internally. 

Best way to measure is to look at `time mc mirror` - i.e time taken for entire data transfer.

**Comment by systemlevel** [comment]
Quick performance snapshot to add to this:

<img width="1176" alt="screen shot 2018-08-28 at 5 37 45 pm" src="https://user-images.githubusercontent.com/894983/44756777-4108d580-aae9-11e8-88b4-3e53da765cda.png">

I will be running some with the `time mc mirror` command as well.
